module/VuFind/src/VuFind/Db/Service/SearchService.php

```diff
@@ -100,15 +105,12 @@ public function createAndPersistEntityWithChecksum(int $checksum): SearchEntityI
      */
     public function destroySession(string $sessionId, UserEntityInterface|int|null $userOrId = null): void
     {
-        $uid = $userOrId instanceof UserEntityInterface ? $userOrId->getId() : $userOrId;
-        $callback = function ($select) use ($sessionId, $uid) {
-            $select->where->equalTo('session_id', $sessionId)->and->equalTo('saved', 0);
-            if ($uid !== null) {
-                $select->where->OR
-                    ->equalTo('user_id', $uid)->and->equalTo('saved', 0);
-            }
-        };
-        $this->getDbTable('search')->delete($callback);
+        $userId = $userOrId instanceof UserEntityInterface ? $userOrId->getId() : $userOrId;
+        $dql = 'DELETE FROM ' . $this->getEntityClass(SearchEntityInterface::class) . ' s '
+        . 'WHERE s.sessionId = :sessionId AND s.saved = 0 AND s.user = :userId';
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameters(compact('sessionId', 'userId'));
```

Comment:
I think we need to retain the original code's special handling of null user ID values to ensure consistent behavior:

```suggestion
        $parameters = compact('sessionId');
        $dql = 'DELETE FROM ' . $this->getEntityClass(SearchEntityInterface::class) . ' s '
            . 'WHERE s.sessionId = :sessionId AND s.saved = 0';
        if ($userId !== null) {
            $dql .= ' AND s.user = :userId';
            $parameters['userId'] = $userId;
        }
        $query = $this->entityManager->createQuery($dql);
        $query->setParameters($parameters);
```